engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java

```diff
@@ -577,4 +572,83 @@ Object getMembers(boolean includeInternal) {
         MethodNames.Module.GET_ASSOCIATED_CONSTRUCTOR,
         MethodNames.Module.EVAL_EXPRESSION);
   }
+
+  private record Sources(
+          TruffleFile sourceFile,
+          AdaptiveRope ropeHolder,
+          Source cachedSource) {
+
+    static final Sources NONE = new Sources(null, new AdaptiveRope(null), null);
+
+    Sources newWith(TruffleFile f) {
+      return new Sources(f, ropeHolder(), cachedSource());
+    }
+
+    Sources newWith(Rope r, boolean keepCachedSource) {
+      if (keepCachedSource) {
+        if (r.characters().length() != literalSource().characters().length()) {
+          throw new IllegalStateException();
+        }
+        this.ropeHolder.update(r);
+        return this;
+      } else {
+        return new Sources(sourceFile(), new AdaptiveRope(r), null);
+      }
+    }
+
+    Sources ensureCachedSource(QualifiedName name) throws IOException {
+      if (cachedSource != null) {
+        return this;
+      }
+      if (literalSource() != null) {
+        var src = Source.newBuilder(LanguageInfo.ID, ropeHolder, name.toString()).build();
+        return new Sources(sourceFile, ropeHolder, src);
+      } else if (sourceFile != null) {
+        var src = Source.newBuilder(LanguageInfo.ID, sourceFile).build();
+        var lit = Rope.apply(src.getCharacters().toString());
+        return new Sources(sourceFile, new AdaptiveRope(lit), src);
+      }
+      throw new IllegalStateException();
+    }
+
+    String getPath() {
+      return sourceFile() == null ? null : sourceFile().getPath();
+    }
+
+    Rope literalSource() {
+      return ropeHolder.rope;
+    }
+
+    private static final class AdaptiveRope implements CharSequence {
```

Comment:
Modifying the `CharSequence`underlaying  a `Source` is a neat trick, but it only works if the length, number of lines doesn't change. If we want to support insertions/removal of characters on a single line, we cannot use the trick as `Source.getTextMap()` (and possibly other invariants) would get broken.

Reply:
No longer in use since [d7deb6c](https://github.com/enso-org/enso/pull/3508/commits/d7deb6c0f9ed4cbf6937282948519d062c72401a)